NCAR/ccpp-framework
# Bug fix for unit conversion error in ccpp_prebuild.py (variables that are slices of a n+1 dimensional array have wrong allocation)

## Description

1. Bug fix in `scripts/mkstatic.py`: define `dim_string_allocate` so that temporary (group cap) variables of rank `n` that correspond to a slice of an `n+1` dimensional array have the right dimensions in the assignment calls and subroutine call lists. See the inline documentation added in `mkstatic.py` around line 1476 for more information.

2. Addition of test `test_prebuild/test_unit_conv` to test for the above and to test for the `capgen` issue reported in https://github.com/NCAR/ccpp-framework/issues/594 (Unit conversion bug when variable is used by two different schemes with different units). Note that `ccpp_prebuild.py` did not suffer from this bug, but I wanted to make sure that we test for it.

3. In the development of the new test `test_unit_conv`, I discovered that we should declare all incoming host variables in the group caps as `target`. This is because it is tricky in prebuild to determine that a parent variable `bar` needs to have the `target` attribute in case a slice of it has the active attribute. This is a bit of an edge case, but I believe this additional attribute is safe. I will run full UFS RTs with this PR to check if the results are b4b or if the addition of `target` causes the compiler to optimize differently.

4. Minor updates to `test_prebuid/{test_blocked_data,test_chunked_data}`: fix wrong name of subroutines in diagnostic error messages, set thread number and thread counter to safe values when running over the entire domain.

User interface changes?: No

Resolves https://github.com/NCAR/ccpp-framework/issues/598

Testing:
  test removed: n/a
  unit tests: docstring tests not run, since the files that are modified in this PR do not have docstring tests
  system tests: CI tests all pass
  manual testing:
    1. Ran `test_prebuild/run_all_tests.sh` on my laptop with `gcc@13.3.0` and `openmpi/5.0.3`
    2. Ran full regression tests with ufs-weather-model on Hera, all tests passed. See https://github.com/ufs-community/ufs-weather-model/pull/2464 and https://github.com/NOAA-EMC/fv3atm/pull/878
